### PR TITLE
Improve build-manifest to work with the improved promote-release

### PR DESCRIPTION
**Pull request by emilyalbini**
This PR makes some changes to build-manifest to have it work better with the other improvements I'm making to [promote-release](https://github.com/rust-lang/promote-release).

A new way to invoke the tool was added: `./x.py run src/tools/build-manifest`. The new invocation disables the generation of `.sha256` files and the generation of GPG signatures, as those steps are not tied to the Rust version we're building the manifest of: handling them in `promote-release` will improve the maintenability of our release process. Invocations through the old command (`./x.py dist hash-and-sign`) are referred inside the source code as "legacy". The new invocation also enables internal parallelism, disabled on legacy to avoid overloading our old server.

Improvements were also made on how the checksums included in the manifest are generated:

* The manifest is first generated with placeholder checksums, and then a function walks through the manifes and calculates only the needed hashes. Before this PR, all the hashes were calculated beforehand, including the hashes of unused files.
* Calculating the hashes is now done in parallel with rayon, to better utilize all the available disk bandwidth.
* The `sha2` crate is now used instead of the `sha256sum` CLI tool: this avoids the overhead of calling another process, but more importantly enables hardware acceleration whenever available (the `sha256sum` CLI tool doesn't support it at all).

r? @Mark-Simulacrum 
This PR is best reviewed commit-by-commit.

### CODE REVIEW

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me with minor nit fixed

File: src/tools/build-manifest/src/main.rs [hunk 1]
```diff
@@ -574,7 +616,16 @@ impl Builder {
...
+            self.hash(&dst);
+            self.sign(&dst);
+        }
     }
 }
+
+fn fetch_hash(path: &Path) -> Result<String, Box<dyn Error>> {
+    let mut file = File::open(path)?;
+    let mut sha256 = sha2::Sha256::default();
+    std::io::copy(&mut file, &mut sha256)?;
```
[the remark this answers is not in the document]
Comment (pietroalbini) [reply src/tools/build-manifest/src/main.rs:637]: Great point!

**Comment by pietroalbini** [comment]
@rustbot modify labels: beta-nominated T-infra

Getting this on beta would reduce the transition period for the new release process.

**Comment by Mark-Simulacrum** [comment]
Just leaving a note here: this is beta-backport for 1.48, not 1.47 which is releasing later this week.

**Comment by pietroalbini** [comment]
The beta backport was approved by the infrastructure team!

@rustbot modify labels: beta-accepted

**Pull request merged**